Cache guards once per variable tracker, rather than re-propagating them repeatedly

mlazos opened the pull request:
This improves tracing performance of optimizer tracing significantly (2x). In essence this just removes the recursion from propagate because it is not necessary. ListVariables and ConstDictVariables already contain the guards from the items contained in them.

Adds two other optimizations for special cases of `recursively_contains`

helps with https://github.com/pytorch/torchdynamo/issues/1803

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

mlazos committed:
Add guard caching and a small rec_contains opt

mlazos committed:
Merge remote-tracking branch 'origin' into mlazos/guard-cache

mlazos committed:
Fix missed mutable check

voznesenskym commented:
Apologies for missing some context... why are we doing this?

voznesenskym commented on torch/_dynamo/variables/base.py:
```diff
@@ -273,8 +278,13 @@ def __init__(
     def __post_init__(self, *args, **kwargs):
+        if self.inner_guards is None:
+            self.inner_guards = set()
+            self.inner_guards = VariableTracker.propagate([self])["guards"]
```
I don't understand why we need to do this? Is it just to call propagation here?

voznesenskym reviewed (commented):
Can you add comments explaining how this works, and maybe a test?

mlazos commented:
> Apologies for missing some context... why are we doing this?

I added more context in the description. Basically the idea is to only gather the guards on a variable tracker once (rather than every time we call `propagate`). I don't plan on adding additional tests for this because it should be covered by existing tests. 

This improves the performance of tracing optimizers because we would constantly be propagating the guards from the same 1000 element list every time we add another parameter to that list (and every time we iterate over it, we split it into another thousand variable trackers where we propagate again).

mlazos committed:
Retriggering ci

mlazos committed:
Revert "Retriggering ci"

This reverts commit 3037b514f64d5e8a189e3f292cb33cc028b64357.

mlazos committed:
Fix bug

jansel reviewed (commented):
Test failures seem real

mlazos committed:
Fix broken test case, and fix bug in propagate.

mlazos committed:
Revert debug changes to test case

mlazos commented:
> Test failures seem real

Agreed, should be fixed now

anijain2305 commented on torch/_dynamo/variables/base.py:
```diff
@@ -265,6 +270,7 @@ def __init__(
         source: Source = None,
         mutable_local: MutableLocal = None,
         recursively_contains: Optional[Set] = None,
+        inner_guards: Optional[Set] = None,
```
Maybe a comment explaining the difference between guards and inner_guards. My understanding is this

guards point to the guards that only cover the surface of the VariableTracker, e.g., for ListVariable, it only contains guards about the list object, but not the items inside the list.


inner_guards is a union of the guards of all the VariableTracker objects used to construct the concerned VariableTracker object.

jansel replied:
Hrm, is this just specific to ListVariable?

Rather than adding new caches, we could just establish the invariant that `list.guards` already contains its recursive guards and not need to add a new property.  There aren't that many places ListVariable is constructed.

mlazos replied:
It's also for dicts, but ideally we should do this for every container, right?

mlazos replied:
I suppose either way `visit` would always need to updated any time a new container is added, so I can just apply this to lists, which is what gives the speedup.

mlazos replied:
Turns out we actually already did this! Both dicts and lists are updated/constructed with the guards contained in them as well.

anijain2305 reviewed (approved):
LGTM. Just a comment suggestion.

mlazos committed:
Remove cache and add invariant for lists, dicts

mlazos committed:
Fix bug

mlazos committed:
Fix list init bug

mlazos committed:
Add contained guards in builder for container types

mlazos committed:
Revert "Add contained guards in builder for container types"

This reverts commit f45c4abd739ee846ff62b278811f16dca84f78a7.

mlazos committed:
Added hint to not regenerate guards to list and dict constrs

mlazos commented:
@pytorchbot merge
